app/assets/javascripts/discourse/app/form-kit/components/fk/form.gjs

```diff
@@ -0,0 +1,340 @@
+          : Reflect.get(target, prop);
+      },
+
+      set(target, property, value) {
+        return Reflect.set(transientData, property, value);
+      },
+
+      has(target, prop) {
+        return prop in transientData ? true : Reflect.has(target, prop);
+      },
+
+      getOwnPropertyDescriptor(target, prop) {
+        return Reflect.getOwnPropertyDescriptor(
+          prop in transientData ? transientData : target,
+          prop
+        );
+      },
+
+      ownKeys(target) {
+        return (
+          [...Reflect.ownKeys(target), ...Reflect.ownKeys(transientData)]
+            // return only unique values
+            .filter((value, index, array) => array.indexOf(value) === index)
+        );
+      },
+
+      deleteProperty(target, prop) {
+        if (prop in transientData) {
+          delete transientData[prop];
+        }
+
+        return true;
+      },
+    });
+  }
+
+  get validateOn() {
+    return this.args.validateOn ?? VALIDATION_TYPES.submit;
+  }
+
+  get fieldValidationEvent() {
+    const { validateOn } = this;
+
+    if (validateOn === VALIDATION_TYPES.submit) {
+      return undefined;
+    }
+
+    return validateOn;
+  }
+
+  get hasErrors() {
+    return Array.from(this.fields.values()).some((field) => field.hasErrors);
+  }
+
+  get errors() {
+    const errors = {};
+    for (const [name, field] of this.fields) {
+      errors[name] = field.errors;
+    }
+    return errors;
+  }
+
+  @action
+  addError(name, message) {
+    const field = this.fields.get(name);
+    field?.pushError(message);
+  }
+
+  @action
+  async set(key, value) {
+    this.isDirtyForm = true;
+    set(this.effectiveData, key, value);
+
+    if (this.fieldValidationEvent === VALIDATION_TYPES.change) {
+      await this.handleFieldValidation(key);
+    }
+
+    await new Promise((resolve) => next(resolve));
+  }
+
+  @action
+  registerField(name, field) {
+    if (!name) {
+      throw new Error("@name is required on `<form.Field />`.");
+    }
+
+    if (this.fields.has(name)) {
+      throw new Error(
+        `@name="${name}", is already in use. Names of \`<form.Field />\` must be unique!`
+      );
+    }
+
+    const fieldModel = new FieldData(name, field);
+    this.fields.set(name, fieldModel);
+
+    return fieldModel;
+  }
+
+  @action
+  unregisterField(name) {
+    this.fields.delete(name);
+  }
+
+  @action
+  async onSubmit(event) {
+    event?.preventDefault();
+
+    await this.validate(this.fields);
+
+    if (!this.hasErrors) {
+      this.isDirtyForm = false;
+      await this.args.onSubmit?.(this.effectiveData);
+    }
+  }
+
+  @action
+  async onReset(event) {
+    event?.preventDefault();
+
+    for (const key of Object.keys(this.transientData)) {
+      delete this.transientData[key];
+    }
+
+    this.fields.forEach((field) => {
+      field.reset();
+    });
+
+    this.fieldsWithErrors = null;
+
+    await this.args.onReset?.(this.effectiveData);
+
+    await new Promise((resolve) => next(resolve));
```

Comment:
I would like to avoid this, but couldn't figure out how to without having to have many `settled()` in my tests